### PR TITLE
46

### PR DESCRIPTION
**Pull request by wentwog**

### CODE REVIEW

**Comment by dmarkov** [comment]
Many thanks, I'll find someone to review it before we merge

**Comment by dmarkov** [comment]
@krzyk please review this PR

File: src/main/java/com/jcabi/log/VerboseProcess.java [hunk 1]
```diff
@@ -266,7 +280,7 @@ private String waitFor() throws InterruptedException {
             Logger.debug(
                 this, "#waitFor(): process finished: %s", this.process
             );
-            if (!done.await(2L, TimeUnit.SECONDS)) {
```
Comment (krzyk) [inline src/main/java/com/jcabi/log/VerboseProcess.java]: this should be left at `2L`

Comment (wentwog) [inline src/main/java/com/jcabi/log/VerboseProcess.java]: only here or everywhere where N_MONITORS constant is introduced ?
2L looks to me as magic number, that's why I've made the replacement.

Comment (krzyk) [inline src/main/java/com/jcabi/log/VerboseProcess.java]: Yes, but 2L here doesn't mean the same as N_MONITORS

File: src/test/java/com/jcabi/log/VerboseProcessTest.java [hunk 2]
```diff
@@ -236,14 +244,163 @@ public void logsErrorWhenUnderlyingStreamIsClosed() throws Exception {
...
+        final StringWriter writer = new StringWriter();
+        final WriterAppender appender = new WriterAppender(
+            new SimpleLayout(),
+            writer
+        );
+        appender.addFilter(new VrbPrcMonitorFilter(verboseProcess));
+        org.apache.log4j.Logger.getLogger(
+            VerboseProcess.class
+        ).addAppender(appender);
+        final int readInterval = 50;
```
Comment (krzyk) [inline src/test/java/com/jcabi/log/VerboseProcessTest.java]: you can use `com.jcabi.aspects.Tv` for numeric constants

File: src/test/java/com/jcabi/log/VerboseProcessTest.java [hunk 3]
```diff
@@ -236,14 +244,163 @@ public void logsErrorWhenUnderlyingStreamIsClosed() throws Exception {
...
+            new TimerTask() {
+                @Override
+                public void run() {
+                    verboseProcess.close();
+                }
+            },
+            readInterval
+        );
+        verboseProcess.stdoutQuietly();
+        final int monitorsInterval = 500;
```
Comment (krzyk) [inline src/test/java/com/jcabi/log/VerboseProcessTest.java]: you can use `com.jcabi.aspects.Tv` for numeric constants, and get rid of this variable

File: src/test/java/com/jcabi/log/VerboseProcessTest.java [hunk 4]
```diff
@@ -236,14 +244,163 @@ public void logsErrorWhenUnderlyingStreamIsClosed() throws Exception {
...
+        }
+
+        @Override
+        public void close() throws IOException {
+            this.closed = true;
+        }
+    }
+
+    /**
+     * Filter of log messages of {@link VerboseProcess}'s monitor threads.<br/>
```
Comment (krzyk) [inline src/test/java/com/jcabi/log/VerboseProcessTest.java]: remove `<br/>` and add <p> on the next line

File: src/test/java/com/jcabi/log/VerboseProcessTest.java [hunk 5]
```diff
@@ -236,14 +244,163 @@ public void logsErrorWhenUnderlyingStreamIsClosed() throws Exception {
...
+         */
+        private static final String THREADNAME_START = "VrbPrc.Monitor-";
+
+        /**
+         * HashCode of {@link VerboseProcess} to filter.
+         */
+        private final transient int hash;
+
+        /**
+         * Create filter for this process.<br/>
```
Comment (krzyk) [inline src/test/java/com/jcabi/log/VerboseProcessTest.java]: remove `<br/>` and add <p> on the next line

File: src/test/java/com/jcabi/log/VerboseProcessTest.java [hunk 6]
```diff
@@ -236,14 +244,163 @@ public void logsErrorWhenUnderlyingStreamIsClosed() throws Exception {
...
+
+        @Override
+        public int read() throws IOException {
+            if (this.closed) {
+                throw new IOException("Stream closed");
+            }
+            final int next;
+            if (this.feed) {
+                this.feed = false;
+                next = LINE_FEED;
```
Comment (krzyk) [inline src/test/java/com/jcabi/log/VerboseProcessTest.java]: Add class name before constant.

File: src/test/java/com/jcabi/log/VerboseProcessTest.java [hunk 7]
```diff
@@ -236,14 +244,163 @@ public void logsErrorWhenUnderlyingStreamIsClosed() throws Exception {
...
+        public VrbPrcMonitorFilter(final VerboseProcess prc) {
+            super();
+            this.hash = prc.hashCode();
+        }
+
+        @Override
+        public int decide(final LoggingEvent event) {
+            final String thread = event.getThreadName();
+            final int decision;
+            if (thread.startsWith(THREADNAME_START)) {
```
Comment (krzyk) [inline src/test/java/com/jcabi/log/VerboseProcessTest.java]: add class name before constant

File: src/main/java/com/jcabi/log/VerboseProcess.java [hunk 8]
```diff
@@ -75,6 +77,11 @@
     private static final String UTF_8 = "UTF-8";
 
     /**
+     * Num of stream monitors.
```
Comment (krzyk) [inline src/main/java/com/jcabi/log/VerboseProcess.java]: Number

**Comment by krzyk** [comment]
@wentwog few comments above

File: src/main/java/com/jcabi/log/VerboseProcess.java (hunk 1 above)
Comment (wentwog) [inline src/main/java/com/jcabi/log/VerboseProcess.java]: I think it means exactly the same: we are waiting for all monitors (N_MONITORS) when they are done.

Comment (krzyk) [inline src/main/java/com/jcabi/log/VerboseProcess.java]: So we are waiting for each monitor for 1 second? Why 1 second per monitor? Previously we had only one monitor and we waited 2 seconds also.

File: src/test/java/com/jcabi/log/VerboseProcessTest.java (hunk 4 above)
Comment (wentwog) [inline src/test/java/com/jcabi/log/VerboseProcessTest.java]: sorry, haven't got. What to add? If just line break then we will not see it in the javadoc.
Afaik, we can leave <br/> or use <p> on the next line to have the break in the javadoc.

File: src/main/java/com/jcabi/log/VerboseProcess.java (hunk 1 above)
Comment (wentwog) [inline src/main/java/com/jcabi/log/VerboseProcess.java]: Ah.. I see, yes the meaning is different, I'll leave `2L`.

File: src/test/java/com/jcabi/log/VerboseProcessTest.java (hunk 4 above)
Comment (wentwog) [inline src/test/java/com/jcabi/log/VerboseProcessTest.java]: I have replaced with `<p>` on the next line, I suppose that was what you've meant.
If not, please, tell me.

Comment (krzyk) [inline src/test/java/com/jcabi/log/VerboseProcessTest.java]: yes, github ate my `<p>` :)

**Comment by wentwog** [comment]
@krzyk I've fixed, please, review.

**Comment by krzyk** [comment]
@rultor merge pls

**Comment by krzyk** [comment]
@wentwog tests are not passing on rultor

**Comment by wentwog** [comment]
@krzyk Yes, investigating...

**Comment by wentwog** [comment]
@krzyk I've fixed, please, review.

File: src/main/java/com/jcabi/log/VerboseProcess.java [hunk 9]
```diff
@@ -176,14 +193,19 @@ public String stdoutQuietly() {
...
-    //  the process. It should be done before destroy() is called. See the
-    //  following for more details: https://github.com/jcabi/jcabi-log/issues/38
-    //  http://www.java2s.com/Code/Java/Threads/Thesafewaytostopathread.htm
     @Override
     public void close() {
+        synchronized (this.monitors) {
+            this.closed = true;
+        }
+        for (final Thread monitor : this.monitors) {
+            if (monitor != null) {
```
Comment (krzyk) [inline src/main/java/com/jcabi/log/VerboseProcess.java]: From what I see it is not possible to have null monitor in this array

Comment (wentwog) [inline src/main/java/com/jcabi/log/VerboseProcess.java]: It is possible. Before `launchMonitors()` is called there will be nulls.

**Comment by krzyk** [comment]
@rultor merge pls

**Comment by dmarkov** [comment]
@krzyk **25 mins** sent to your balance (ID `50572986`), many thanks!... added +25 to your rating, now it is equal to [+2235](http://www.netbout.com/b/35035?open=rating)

**Comment by dmarkov** [comment]
@rultor deploy pls

**Pull request merged**